backstory/src/main/java/com/google/sps/story/data/StoryEndingTools.java

```diff
@@ -0,0 +1,106 @@
+
+package com.google.sps.story.data;
+
+/**
+ * Static methods to help end a story naturally
+ */
+public class StoryEndingTools {
+  /** the endings to add to the story (should be one sentence) */
+  private static String[] ENDINGS = {
+      "The End.",
+      "They lived happily ever after.",
+      "Then, everything went horribly wrong.",
+      "And--- that's a wrap.",
+      "Goodbye!",
+      "Then the director screamed \"CUT!\".",
+      "With that, our story draws to a close.",
+      "We'll never know what happened next.",
+  };
+
+  /**
+   * Overrides default constructor to ensure class can't be instantiated.
+   */
+  private StoryEndingTools() {
+    throw new AssertionError();
+  }
+
+  /**
+   * Takes a story from GPT-2 and returns onee with a more natural ending.
+   *
+   * @param story the story to return with a natural ending
+   * @return story with ending
+   * @throws IllegalArgumentException if story is null
+   */
+  public static String endStory(String story) {
+    if (story == null) {
+      throw new IllegalArgumentException("Story should not be null");
+    }
+
+    story = removeSentenceFragmentAtEnd(story);
+    story = addEnding(story);
+
+    return story;
+  }
+
+  /**
+   * Remove a sentence fragment at the end of the story (if there is one)
+   * as that commonly occurs with GPT2-generated text. Returns the story
+   * up to the last sentence-ending punctuation (period, exclamation point,
+   * or question mark).
+   *
+   * @param story to remove the sentence fragment from
+   * @return the passed-in story without the last sentence fragment
+   * @throws IllegalArgumentException if story is null
+   */
+  public static String removeSentenceFragmentAtEnd(String story) {
+    if (story == null) {
+      throw new IllegalArgumentException("Story should not be null");
+    }
+
+    int lastPeriodIndex = story.lastIndexOf(".");
+    int lastExclamationPointIndex = story.lastIndexOf("!");
+    int lastQuestionMarkIndex = story.lastIndexOf("?");
+
+    // find the last sentence-ending punctuation by finding the max number
+    // amongst all the sentence-ending types indices
+    int lastSentenceEnder = Math.max(lastPeriodIndex, lastExclamationPointIndex);
+    lastSentenceEnder = Math.max(lastSentenceEnder, lastQuestionMarkIndex);
+
+    // if last period coincides with end of story, then return story as is
+    if (lastSentenceEnder == story.length() - 1) {
+      return story;
+    }
+
+    // else, chop off story after last period
+    return story.substring(0, lastSentenceEnder + 1);
+  }
+
+  /**
+   * Add one of a list of pre-written endings to the story, and return
+   * the story with that ending.
+   *
+   * @param story the story to add an ending to
+   * @return story + ending
+   * @throws IllegalArgumentException if story is null
+   */
+  public static String addEnding(String story) {
+    if (story == null) {
+      throw new IllegalArgumentException("Story should not be null");
+    }
```

Comment:
Also check length of story after trimming whitespace?

Reply:
Similarly to above, confused as to what trimming whitespace refers to.

Reply:
I'll make the change you suggested above here as well.

Reply:
Done